plugins/glustershd/transactions.go

```diff
@@ -0,0 +1,106 @@
+package glustershd
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/gluster/glusterd2/glusterd2/brick"
+	"github.com/gluster/glusterd2/glusterd2/daemon"
+	"github.com/gluster/glusterd2/glusterd2/servers/sunrpc/dict"
+	"github.com/gluster/glusterd2/glusterd2/transaction"
+	"github.com/gluster/glusterd2/glusterd2/volume"
+)
+
+func getHxlChildrenCount(volinfo *volume.Volinfo) (int, string) {
+	if volinfo.Type == volume.Replicate || volinfo.Type == volume.DistReplicate {
+		return volinfo.Subvols[0].ReplicaCount, "replicate"
+	}
+	return volinfo.Subvols[0].DisperseCount, "disperse"
+}
+
+func addHxlatorToDict(reqDict map[string]string, volinfo *volume.Volinfo, index int, count int, xlType string) map[string]string {
+	key := fmt.Sprintf("xl-%d", count)
+	xName := fmt.Sprintf("%s-%s-%d", volinfo.Name, xlType, index)
+	reqDict[key] = xName
+	reqDict[xName] = strconv.Itoa(index)
+	return reqDict
+}
+
+func selectHxlatorsWithBricks(volinfo *volume.Volinfo, healType string) map[string]string {
+	index := 1
+	hxlatorCount := 0
+	reqDict := make(map[string]string)
+	reqDict["heal-op"] = healType
+	reqDict["xl-op"] = reqDict["heal-op"]
+	reqDict["volname"] = volinfo.Name
+	reqDict["sync-mgmt-operation"] = strconv.Itoa(20)
+	reqDict["vol-id"] = volinfo.ID.String()
+	hxlChildren, xlType := getHxlChildrenCount(volinfo)
+	volBricks := volinfo.GetBricks()
+	for brick := range volBricks {
+		hostKey := fmt.Sprintf("%d-hostname", index-1)
+		brickPathKey := fmt.Sprintf("%d-path", index-1)
+		reqDict[hostKey] = volBricks[brick].Hostname
+		reqDict[brickPathKey] = volBricks[brick].Path
+
+		if index%hxlChildren == 0 {
+			reqDict = addHxlatorToDict(reqDict, volinfo, (index-1)/hxlChildren, hxlatorCount, xlType)
+			hxlatorCount++
+		}
+		index++
+	}
+	reqDict["count"] = strconv.Itoa(hxlatorCount)
+	return reqDict
+}
```

Comment:
@itisravi Requesting you to review the arguments being send in the RPC call. The above 3 functions in particular need your review.

Reply:
@vpandey-RH  I'm not quite able to map this change with gd1. In gd1, when you launch index heal, each node's glusterd will send the xlator-op to its respective glustershd. The dict sent to each shd will contain 'count' no. of translators. The names of the translators will be the afr xlators present on that node. glusterd1 uses gf_uuid_compare() to decide the xlators to be added to the dict.

Reply:
Example:
>
> Volume Name: testvol
Type: Distributed-Replicate
Volume ID: ec9e1c13-b4b0-4024-a49e-e3a3141a29be
Status: Started
Snapshot Count: 0
Number of Bricks: 5 x 3 = 15
Transport-type: tcp
Bricks:
Brick1:  N1:/bricks/brick1
Brick2:  N2:/bricks/brick1
Brick3:  N3:/bricks/brick1

> Brick4:  N4:/bricks/brick2
Brick5:  N5:/bricks/brick2
Brick6:  N6:/bricks/brick2

> Brick7:  N1:/bricks/brick3
Brick8:  N2:/bricks/brick3
Brick9:  N3:/bricks/brick3

> Brick10: N4:/bricks/brick4
Brick11: N5:/bricks/brick4
Brick12: N6:/bricks/brick4

> Brick13: N1:/bricks/brick5
Brick14: N2:/bricks/brick5
Brick15: N3:/bricks/brick5


For example, in the above volume,
glusterd on say, N1 must send 3 items (ie. count = 3) in the dict:
xl-0:testvol-replicate-0
xl-0:testvol-replicate-2
xl-0:testvol-replicate-4

Similarly, glusterd on say, N6 must send 2 items (ie. count = 2) in the dict:
xl-0:testvol-replicate-1
xl-0:testvol-replicate-3